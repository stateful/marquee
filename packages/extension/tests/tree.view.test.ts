import { TreeView } from '../src/tree.view'

const context = {} as any
const stateMgr = {
  todoWidget: { on: jest.fn () },
  snippetsWidget: { on: jest.fn () },
  notesWidget: { on: jest.fn () },
  global: { on: jest.fn (), updateState: jest.fn(), state: { globalScope: true } },
  projectWidget: { getActiveWorkspace: jest.fn().mockReturnValue({ id: '123' })}
} as any

beforeEach(() => {
  context.subscriptions = []
  context.globalState = { get: jest.fn((key: string, defaultParam: any) => defaultParam) }
  context.extensionUri = '/foo/bar'
  stateMgr.handleUpdates = jest.fn()
  stateMgr.recover = jest.fn().mockReturnValue({
    globalScope: true
  })
  stateMgr.getActiveWorkspace = jest.fn()
  stateMgr.save = jest.fn()
})

describe('TreeView', () => {
  test('constructor', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    expect(context.subscriptions).toEqual(['marquee.toggleScope'])
    expect(tree['update']).toBeCalledTimes(1)
  })

  test('clearTree', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    tree.clearTree()
    expect(tree['state']).toMatchSnapshot()
  })

  test('toggleScope', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    tree.toggleScope()
    expect(tree['stateMgr'].global.updateState).toBeCalledWith('globalScope', false)
    expect(tree['update']).toBeCalledTimes(2)
  })

  test('_updateTodos', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    tree['_updateTodos']({} as any, true)
    expect(tree['state'].todos).toHaveLength(0)

    context.globalState.get.mockReturnValue({
      todos: [{ workspaceId: '123' }, { workspaceId: '321' }]
    })
    tree['_updateTodos']({ id: '321' } as any, true)
    expect(tree['toplevel']).toMatchSnapshot()
    expect(tree['state'].todos).toMatchSnapshot()
  })

  test('_updateNotes', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    tree['_updateNotes']({} as any, true)
    expect(tree['state'].notes).toHaveLength(0)

    context.globalState.get.mockReturnValue({
      notes: [{ workspaceId: '123' }, { workspaceId: '321' }]
    })
    tree['_updateNotes']({ id: '321' } as any, true)
    expect(tree['toplevel']).toMatchSnapshot()
    expect(tree['state'].notes).toMatchSnapshot()
  })

  test('_updateSnippets', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    tree['_updateSnippets']({} as any, true)
    expect(tree['state'].snippets).toHaveLength(0)

    context.globalState.get.mockReturnValue({
      snippets: [{ workspaceId: '123' }, { workspaceId: '321' }]
    })
    tree['_updateSnippets']({ id: '321' } as any, true)
    expect(tree['toplevel']).toMatchSnapshot()
    expect(tree['state'].snippets).toMatchSnapshot()
  })

  test('update', () => {
    jest.spyOn(TreeView.prototype, 'refresh' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    context.globalState.get.mockReturnValue({
      globalScope: true
    })
    tree['_updateTodos'] = jest.fn()
    tree['_updateNotes'] = jest.fn()
    tree['_updateSnippets'] = jest.fn()
    tree['refresh'] = jest.fn()

    tree['update']()
    expect(tree['_updateTodos']).toBeCalledWith({ id: '123' }, true)
    expect(tree['_updateNotes']).toBeCalledWith({ id: '123' }, true)
    expect(tree['_updateSnippets']).toBeCalledWith({ id: '123' }, true)
    expect(tree['refresh']).toBeCalledTimes(1)
  })

  test('refresh', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    // @ts-expect-error
    tree['_onDidChangeTreeData'] = { fire: jest.fn() } as any
    tree.refresh()
    expect(tree['_onDidChangeTreeData'].fire).toBeCalledWith(undefined)
  })

  test('getChildren', async () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    tree['refresh'] = jest.fn()
    tree['update']()

    const elems = await tree.getChildren()
    expect(elems).toMatchSnapshot()

    const todos = await tree.getChildren(elems[0])
    expect(todos).toMatchSnapshot()

    const snippets = await tree.getChildren(elems[1])
    expect(snippets).toMatchSnapshot()

    const notes = await tree.getChildren(elems[2])
    expect(notes).toMatchSnapshot()
  })

  test('getTreeItem', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    expect(tree.getTreeItem('foo' as any)).toBe('foo')
  })

  test('getParent', () => {
    jest.spyOn(TreeView.prototype, 'update' as any)
      .mockImplementation(() => ({} as any))
    const tree = new TreeView(context as any, stateMgr)
    expect(tree.getParent()).toBe(null)
  })
})