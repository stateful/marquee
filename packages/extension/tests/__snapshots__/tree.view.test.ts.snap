// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView _updateNotes 1`] = `
Array [
  Object {
    "caption": "Todos",
    "id": 0,
    "type": "todos",
  },
  Object {
    "caption": "Snippets",
    "id": 1,
    "type": "snippets",
  },
  Object {
    "caption": "Notes",
    "id": 2,
    "type": "notes",
  },
]
`;

exports[`TreeView _updateNotes 2`] = `
Array [
  Object {
    "workspaceId": "123",
  },
  Object {
    "workspaceId": "321",
  },
]
`;

exports[`TreeView _updateSnippets 1`] = `
Array [
  Object {
    "caption": "Todos",
    "id": 0,
    "type": "todos",
  },
  Object {
    "caption": "Snippets",
    "id": 1,
    "type": "snippets",
  },
  Object {
    "caption": "Notes",
    "id": 2,
    "type": "notes",
  },
]
`;

exports[`TreeView _updateSnippets 2`] = `
Array [
  Object {
    "workspaceId": "123",
  },
  Object {
    "workspaceId": "321",
  },
]
`;

exports[`TreeView _updateTodos 1`] = `
Array [
  Object {
    "caption": "Todo [global] (0 open / 2 closed)",
    "id": 0,
    "type": "todos",
  },
  Object {
    "caption": "Snippets",
    "id": 1,
    "type": "snippets",
  },
  Object {
    "caption": "Notes",
    "id": 2,
    "type": "notes",
  },
]
`;

exports[`TreeView _updateTodos 2`] = `
Array [
  Object {
    "workspaceId": "123",
  },
  Object {
    "workspaceId": "321",
  },
]
`;

exports[`TreeView clearTree 1`] = `
Object {
  "notes": Array [],
  "snippets": Array [],
  "todos": Array [],
}
`;

exports[`TreeView getChildren 1`] = `
Array [
  Item {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": "foo",
    "command": undefined,
    "contextValue": "todosHeadline",
    "id": "0",
    "isTreeItem": true,
    "label": "Todos",
    "type": "todos",
  },
  Item {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": "foo",
    "command": undefined,
    "contextValue": "snippetsHeadline",
    "id": "1",
    "isTreeItem": true,
    "label": "Snippets",
    "type": "snippets",
  },
  Item {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": "foo",
    "command": undefined,
    "contextValue": "notesHeadline",
    "id": "2",
    "isTreeItem": true,
    "label": "Notes",
    "type": "notes",
  },
]
`;

exports[`TreeView getChildren 2`] = `
Array [
  TodoItem {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": undefined,
    "command": Object {
      "command": "marquee.todo.addEmpty",
      "title": "Add new todo",
    },
    "contextValue": "AddNew",
    "iconPath": Object {
      "dark": "/foo/bar/assets/add-light.svg",
      "light": "/foo/bar/assets/add-dark.svg",
    },
    "id": "addtodos",
    "isTreeItem": true,
    "item": Object {},
    "label": "Add new todo",
    "type": "AddNew",
  },
]
`;

exports[`TreeView getChildren 3`] = `
Array [
  SnippetItem {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": undefined,
    "command": Object {
      "arguments": Array [
        Object {
          "workspaceId": "123",
        },
      ],
      "command": "marquee.snippet.insert",
      "title": "Insert Snippet",
    },
    "contextValue": "Snippet",
    "empty": false,
    "iconPath": Object {
      "dark": "/foo/bar/assets/snippets-light.svg",
      "light": "/foo/bar/assets/snippets-dark.svg",
    },
    "id": undefined,
    "isTreeItem": true,
    "item": Object {
      "workspaceId": "123",
    },
    "label": undefined,
    "type": "Snippet",
  },
  SnippetItem {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": undefined,
    "command": Object {
      "arguments": Array [
        Object {
          "workspaceId": "321",
        },
      ],
      "command": "marquee.snippet.insert",
      "title": "Insert Snippet",
    },
    "contextValue": "Snippet",
    "empty": false,
    "iconPath": Object {
      "dark": "/foo/bar/assets/snippets-light.svg",
      "light": "/foo/bar/assets/snippets-dark.svg",
    },
    "id": undefined,
    "isTreeItem": true,
    "item": Object {
      "workspaceId": "321",
    },
    "label": undefined,
    "type": "Snippet",
  },
]
`;

exports[`TreeView getChildren 4`] = `
Array [
  NoteItem {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": undefined,
    "command": Object {
      "arguments": Array [
        [Circular],
      ],
      "command": "marquee.edit",
      "title": "Open note",
    },
    "contextValue": "Note",
    "empty": false,
    "iconPath": Object {
      "dark": "/foo/bar/assets/notes-light.svg",
      "light": "/foo/bar/assets/notes-dark.svg",
    },
    "id": undefined,
    "isTreeItem": true,
    "item": Object {
      "workspaceId": "123",
    },
    "label": undefined,
    "type": "Note",
  },
  NoteItem {
    "archived": undefined,
    "basePath": "/foo/bar",
    "checked": undefined,
    "collapsibleState": undefined,
    "command": Object {
      "arguments": Array [
        [Circular],
      ],
      "command": "marquee.edit",
      "title": "Open note",
    },
    "contextValue": "Note",
    "empty": false,
    "iconPath": Object {
      "dark": "/foo/bar/assets/notes-light.svg",
      "light": "/foo/bar/assets/notes-dark.svg",
    },
    "id": undefined,
    "isTreeItem": true,
    "item": Object {
      "workspaceId": "321",
    },
    "label": undefined,
    "type": "Note",
  },
]
`;
